tests/melt_visco_plastic.cc

```diff
@@ -0,0 +1,1048 @@
+              if (viscous_stress >= yield_strength)
+                {
+                  plastic_yielding = true;
+                  out.viscosities[i] = yield_strength / (2.0 * edot_ii);
+                }
+
+              // Calculate average internal friction angle and cohesion values
+              const double cohesion = MaterialUtilities::average_value(volume_fractions, cohesions, viscosity_averaging);
+              const double angle_internal_friction = MaterialUtilities::average_value(volume_fractions, angles_internal_friction, viscosity_averaging);
+
+              PlasticAdditionalOutputs<dim> *plastic_out = out.template get_additional_output<PlasticAdditionalOutputs<dim> >();
+              if (plastic_out != nullptr)
+                {
+                  plastic_out->cohesions[i] = cohesion;
+                  plastic_out->friction_angles[i] = angle_internal_friction;
+                  plastic_out->yielding[i] = plastic_yielding ? 1 : 0;
+                }
+
+              // Limit the viscosity with specified minimum and maximum bounds
+              out.viscosities[i] = std::min(std::max(out.viscosities[i], min_viscosity), max_viscosity);
+
+              // Compute the volumetric yield strength (Keller et al. eq (38))
+              volumetric_yield_strength[i] = viscous_stress - tensile_strength;
+            }
+        }
+
+      // fill melt outputs if they exist
+      MeltOutputs<dim> *melt_out = out.template get_additional_output<MeltOutputs<dim> >();
+
+      if (melt_out != NULL)
+        {
+          for (unsigned int i=0; i<in.position.size(); ++i)
+            {
+              const unsigned int porosity_idx = this->introspection().compositional_index_for_name("porosity");
+              double porosity = std::min(1.0, std::max(in.composition[i][porosity_idx],0.0));
+              melt_out->fluid_viscosities[i] = eta_f;
+              melt_out->permeabilities[i] = reference_permeability * std::pow(porosity,3) * std::pow(1.0-porosity,2);
+
+              melt_out->fluid_densities[i] = out.densities[i] + melt_density_change;
+              melt_out->fluid_density_gradients[i] = 0.0;
+
+              const double compaction_pressure = (1.0 - porosity) * (in.pressure[i] - fluid_pressures[i]);
+
+              const double phi_0 = 0.05;
+              porosity = std::max(std::min(porosity,0.995),1.e-8);
+              // compaction viscosities (Keller et al. eq (51)
+              melt_out->compaction_viscosities[i] = intrinsic_viscosities[i] * phi_0 / porosity;
+
+              // visco(elastic) compaction viscosity
+              // Keller et al. eq (36)
+              // TODO include elastic part
+              melt_out->compaction_viscosities[i] *= (1. - porosity);
+
+              // TODO compaction stress evolution parameter
+              // Keller et al. eq. (41) and (44)
+
+              // effective compaction viscosity (Keller et al. eq (43) )
+              // NB: I've added a minus sign as according to eq 43
```

Comment:
From looking at eq 43, should the larger-than sign in the paper not be smaller-than? So P_c<P_y? If so, I'd add that here.

Reply:
Good catch. That would match what the text describes.

@jdannberg Is there something else going on here? If it's just a typo in equation 43 I'll switch the conditional on line 631 here to fix it.

Reply:
I don't think it's a typo in the paper, because it's consistent with equation (38) and the description in the text. But it's worth checking if we have all the signs defined in the same way as in the paper.